### ISSUE TITLE
Provide ability to set annotations on worker nodes

### ISSUE THREAD

**Issue by shyamradhakrishnan** [kubernetes-sigs/cluster-api#7409]
**User Story**

As an operator I would like to have the ability to set annotations on worker nodes created by CAPI. This is similar to the ability to set labels which is being currently worked on.

**Detailed Description**

In many cases, annotations are used to set metadata on nodes. For example in OCI, compartment id can be set on nodes which will help in discovery of nodes by cloud provider, as well as lookup vnics. Other providers may also have such use cases.

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature

**Comment by sbueringer** [comment]
> For example in OCI, compartment id can be set on nodes which will help in discovery of nodes by cloud provider, as well as lookup vnics.

I'm fine in general with the requirement. Just about this one. I don't exactly know what a compartment is, but I wonder if it's the kind of metadata that in some clouds is provided by the cloud-provider running inside of the workload cluster.

**Comment by fabriziopandini** [comment]
/triage accepted
/help

I think we should implement label propagation first, and then build on top of it

**Comment by kfox1111** [comment]
taints is important too. should that be another issue too or wrapped up with labels or annotations?

**Comment by musaprg** [comment]
Label propagation has been supported by https://github.com/kubernetes-sigs/cluster-api/pull/7173. It would be a good time to step onto annotations. (AFAIK it's not implemented for now)

We may have two questions that need to have consensus:

Q1. Should we only allow annotations with predefined prefixes?
Q2. Which annotation value should have precedence if the specified one on the Machine object exists on the corresponding Node object?

For labels, CAPI only allows propagating [managed labels](https://github.com/kubernetes-sigs/cluster-api/blob/c7b145049a4a3649e547a18b61c40f0bce5ddde0/internal/controllers/machine/machine_controller_noderef.go#L163-L179). To achieve the original motivation of this PR, I suppose we anyway need to allow arbitrary annotations on Machine objects to be propagated.

In terms of Q2, for labels, ones on Machine objects always have precedence. IMO it would make sense to follow the same policy for annotations as well.

I'm interested in working on implementation and ready to prepare a PR.

**Comment by nrb** [comment]
https://github.com/kubernetes-sigs/cluster-api/issues/7731 has been finished for quite some time, and https://github.com/kubernetes-sigs/cluster-api/pull/11650 is currently in flight to sync labels from Machines to Nodes.

@fabriziopandini @sbueringer I understand your time is limited - what would you like to see in order to move this forward?

I think your questions are good & warrant some discussion, @musaprg. Do you still have bandwidth to collaborate on this?

**Comment by musaprg** [comment]
Yes, I'm still keen to work on this.

I've looked through https://github.com/kubernetes-sigs/cluster-api/pull/11650. We could take the same approach for propagating annotations.

**Comment by blind3dd** [comment]
is this regarding CAPI feature that would inject annotations?

**Comment by nrb** [comment]
@sbueringer @fabriziopandini @musaprg 

I've taken an initial pass at some code, but before I go further, I want to check with folks to get on the same page.

Currently, CAPI handles a specific set known of labels; anything outside of the set is ignored until #11650 merges. Even after that PR merges, users can still provide a set of known domains to sync, rather than any and all labels found on the `Machine`.

For annotations, right now the only annotations that are copied to a `Machine` are those defined in the `MachineSet.Spec.Template.Annotations` field ([code](https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/machineset/machineset_controller.go#L998)).

When syncing a `Machine`'s metadata to its `Node`, CAPI sets a small set of annotations for its own bookkeeping ([code](https://github.com/kubernetes-sigs/cluster-api/tree/main/internal/controllers#L117)), but otherwise all labels are unaffected.

This leads to the questions from above, which we'll need to answer; maybe they should be in a CAEP?

> Q1. Should we only allow annotations with predefined prefixes?

From my perspective, it should mirror the behavior that #11650 establishes for labels. I'll leave out the specific implementation details since it's not merged yet, but I think it would be best to treat labels and annotations as similarly as possible.

> Q2. Which annotation value should have precedence if the specified one on the Machine object exists on the corresponding Node object?

This one's trickier. Should we perhaps make use of the SSA owner logic, perhaps? My understanding is that labels propagated from CAPI objects to their dependents will have the field's manager set to the relevant controller ([`MachineSet` example](https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/machineset/machineset_controller.go#L560-L590))?

If we're able to say that annotations originating from the `Machine` are owned by CAPI, we could then leave any other annotations from external systems alone and avoid the precedence decision.

If we _can't_ do that, I'm inclined to say that the `Machine`'s value should win if there's a conflict.

**Comment by sbueringer** [comment]
> is this regarding potential CAPI feature that would inject annotations for instance with webhooks?

No we would not inject annotation with a webhook

**Comment by sbueringer** [comment]
@nrb 

> From my perspective, it should mirror the behavior that https://github.com/kubernetes-sigs/cluster-api/pull/11650 establishes for labels. I'll leave out the specific implementation details since it's not merged yet, but I think it would be best to treat labels and annotations as similarly as possible.

tl;dr yes, details:

I would also treat annotations in a very similar way to labels.

For labels we ended up with syncing roughly the following labels per default (I"m ignoring the nuances/differences on how we match each of these):
* node-role.kubernetes.io
* node-restriction.kubernetes.io
* node.cluster.x-k8s.io

The first two are some pre-defined labels from upstream k/k.

I think we should only sync `node.cluster.x-k8s.io` annotations per default.

Like for labels, I think we should add a flag that allows to specify additional annotations to sync.

> This leads to the questions from above, which we'll need to answer; maybe they should be in a CAEP?

Probably it makes sense to generalize the "Label Sync Between Machines and underlying Kubernetes Nodes" proposal to also include annotations (compared to copy&pasting most of the proposal). Would be good to get @fabriziopandini's take on that though.

> Q2. Which annotation value should have precedence if the specified one on the Machine object exists on the corresponding Node object?

I would implement it the exact same way as we did for labels. We tried to use SSA at the time, but then realized it doesn't work: https://github.com/kubernetes-sigs/cluster-api/issues/8417.

IIRC this comes down to that the annotations on the Machine take precedence, and I think that's okay (after all we're restricting the annotations that are synced, we're not just syncing everything)

**Comment by fabriziopandini** [comment]
I agree with @sbueringer 
Might be worth clarify a little bit two points: 

CAPI is authoritative for labels it syncs, there is no co-ownership or conflicts like in SSA (see [code](https://github.com/kubernetes-sigs/cluster-api/blob/9f784575ff59f546fce3c530b10ab2d52b2fa184/internal/controllers/machine/machine_controller_noderef.go#L283-L309)). That means that e.g for the node.cluster.x-k8s.io domain:
- if someone removes from the node a label (set by CAPI) in this domain, CAPI will set it back 
- if someone changes the value for a label in this domain, CAPI will revert this change 
- If CAPI removes a label in this domain, the label will go away
I think we should do the same for annotations

I hope we should only implement propagation of annotation from Machine to Node (extending this [proposal](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220927-label-sync-between-machine-and-nodes.md)), and keep propagation of metadata across CAPI object as it is today (this [proposal](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20221003-In-place-propagation-of-Kubernetes-objects-only-changes.md))

Last but not least, a reminder that above proposals are not yet implemented for MachinePools

**Comment by nrb** [comment]
@sbueringer @fabriziopandini @musaprg I put up https://github.com/kubernetes-sigs/cluster-api/pull/11813, PTAL when you can.
